### PR TITLE
Use upb_proto_library rule to dynamically generate upb codegen in bazel builds

### PR DESCRIPTION
**Pull request by markdroth**

### CODE REVIEW

**Comment by markdroth** [comment]
I've updated this to merge in the changes from when we did the upb upgrade and started using the upbdefs files.  The project is temporarily on hold until we can resolve some internal dependency issues.

The main thing that we still need to resolve here is to fix the non-bazel builds.  We can pick that up again when we're ready to proceed.

**Comment by jtattermusch** [comment]
I wanted to check whether this PR would be compatible with splitting  upb generated filed into `upb-generated` and `upbdefs-generated` directories (see https://github.com/grpc/grpc/pull/23835), but to be honest the ideas in this PR are very hard to follow when description of this PR is empty.
It would be good if there was a short list of things that this PR plans to do - I understand that for bazel build we probably want to start relying on upb_proto_library rules, but it's unclear to me how codegen and source file includes is intended to be handled for other build systems (do we still plan to check in upb generated files? - I guess we have to).

So far I haven't seen anything that would categorically prevent changes from  https://github.com/grpc/grpc/pull/23835 from working with this PR (with some adjustments) but as I said I don't the ideas behind this PR 100% so it's hard to tell.
